staging/src/k8s.io/apiserver/pkg/util/x509metrics/missing_san.go

```diff
@@ -0,0 +1,83 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package x509metrics
+
+import (
+	"crypto/x509"
+	"errors"
+	"net/http"
+	"strings"
+
+	"k8s.io/component-base/metrics"
+)
+
+type x509MissingSANErrorMetricsRTWrapper struct {
+	rt http.RoundTripper
+
+	counter *metrics.CounterVec
+}
+
+// MissingSANRoundTripperWrapper returns a RoundTripper wrapper that's usable
+// within ClientConfig.Wrap that increases the `metricCounter` whenever:
+// 1. we get a x509.HostnameError with string `x509: certificate relies on legacy Common Name field`
+//    which indicates an error caused by the deprecation of Common Name field when veryfing remote
+//    hostname
+// 2. the server certificate in response contains no SAN. This indicates that this binary was
+//    compiled with the GODEBUG=x509ignoreCN=0 in env
+func MissingSANRoundTripperWrapper(metricCounter *metrics.CounterVec) func(rt http.RoundTripper) http.RoundTripper {
+	w := x509MissingSANErrorMetricsRTWrapper{
+		counter: metricCounter,
+	}
+	return w.withRoundTripper
+}
+
+func (w *x509MissingSANErrorMetricsRTWrapper) withRoundTripper(rt http.RoundTripper) http.RoundTripper {
+	w.rt = rt
+	return w
+}
+
+func (w *x509MissingSANErrorMetricsRTWrapper) RoundTrip(req *http.Request) (*http.Response, error) {
```

Comment:
this should implement net.RoundTripperWrapper so internal dialers/tls config can be extracted (so DialerFor works)

Reply:
done